File: level_1/turtlesim/ros2_ws/src/turtle_py/turtle_py/distance_monitor.py
import rclpy
from rclpy.node import Node
from rclpy.qos import QoSProfile
from std_msgs.msg import Float32

class DistanceMonitor(Node):
    def __init__(self):
        super().__init__('distance_monitor')
        self.declare_parameter('warn_distance', 2.5)
        self.subscription = self.create_subscription(Float32, 'turtle_distance', self.callback, QoSProfile(depth=10))
        self.get_logger().info('monitor node up')
    def callback(self, msg):
        limit = float(self.get_parameter('warn_distance').value)
        if msg.data > limit:
            self.get_logger().warning(f'distance {msg.data:.3f} exceeds {limit:.3f}')

def main(args=None):
    rclpy.init(args=args); node = DistanceMonitor()
    try: rclpy.spin(node)
    except KeyboardInterrupt: pass
    finally: node.destroy_node(); rclpy.shutdown() if rclpy.ok() else None

File: level_1/turtlesim/ros2_ws/src/turtle_cpp/src/distance_monitor.cpp
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/float32.hpp"
class DistanceMonitor : public rclcpp::Node { public: DistanceMonitor():Node("distance_monitor_cpp"){ sub_=create_subscription<std_msgs::msg::Float32>("/turtle_distance",10,[this](const auto m){RCLCPP_INFO(get_logger(),"distance=%.3f",m->data);}); RCLCPP_INFO(get_logger(),"distance C++ monitor started"); } private:rclcpp::Subscription<std_msgs::msg::Float32>::SharedPtr sub_;};
int main(int argc,char** argv){rclcpp::init(argc,argv);rclcpp::spin(std::make_shared<DistanceMonitor>());rclcpp::shutdown();}
